lib/host/include/Memory.hpp

```diff
@@ -0,0 +1,145 @@
+//******************************************************************************
+// Copyright (c) 2018, The Regents of the University of California (Regents).
+// All Rights Reserved. See LICENSE for license details.
+//------------------------------------------------------------------------------
+#pragma once
+
+#include <assert.h>
+#include <fcntl.h>
+#include <stdarg.h>
+#include <stddef.h>
+#include <sys/ioctl.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <cerrno>
+#include <cstring>
+#include <iostream>
+#include "./common.h"
+#include "KeystoneDevice.hpp"
+#include "hash_util.hpp"
+
+/*
+ * These are used to make use of C type-checking..
+ */
+typedef struct { uintptr_t pte; } pte_t;
+
+#define pte_val(x) ((x).pte)
+
+#define __pa(x) ((uintptr_t)(x))
+
+#define __pte(x) ((pte_t){(x)})
+
+// page table entry (PTE) fields
+#define PTE_V 0x001     // Valid
+#define PTE_R 0x002     // Read
+#define PTE_W 0x004     // Write
+#define PTE_X 0x008     // Execute
+#define PTE_U 0x010     // User
+#define PTE_G 0x020     // Global
+#define PTE_A 0x040     // Accessed
+#define PTE_D 0x080     // Dirty
+#define PTE_SOFT 0x300  // Reserved for Software
+
+#define PTE_PPN_SHIFT 10
+
+#define VA_BITS 39
+
+#define RISCV_PGLEVEL_BITS 9
+#define RISCV_PGSHIFT 12
+#define RISCV_PGSIZE (1 << RISCV_PGSHIFT)
+
+#if __riscv_xlen == 64
+#define RISCV_PGLEVEL_MASK 0x1ff
+#define RISCV_PGTABLE_HIGHEST_BIT 0x100
+#else
+#define RISCV_PGLEVEL_MASK 0x3ff
+#define RISCV_PGTABLE_HIGHEST_BIT 0x300
+#endif
+
+#define RISCV_PGLEVEL_TOP ((VA_BITS - RISCV_PGSHIFT) / RISCV_PGLEVEL_BITS)
+
+class Memory {
```

Comment:
Yeah, that's necessary. I guess the right way is to put everything into a keystone namespace instead of changing the class names. I'll work on that if that sounds ok